saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticatedPrincipal.java

```diff
@@ -16,13 +16,39 @@
 
 package org.springframework.security.saml2.provider.service.authentication;
 
+import org.springframework.lang.Nullable;
 import org.springframework.security.core.AuthenticatedPrincipal;
 
+import java.util.Collections;
+import java.util.Map;
+
 /**
  * Saml2 representation of an {@link AuthenticatedPrincipal}.
  *
  * @author Clement Stoquart
  * @since 5.2.2
  */
 public interface Saml2AuthenticatedPrincipal extends AuthenticatedPrincipal {
+	/**
+	 * Get the Saml2 token attribute by name
+	 *
+	 * @param name the name of the attribute
+	 * @param <A> the type of the attribute
+	 * @return the attribute or {@code null} otherwise
+	 * @since 5.4
+	 */
+	@Nullable
+	default <A> A getAttribute(String name) {
```

Comment:
Yeah, I prefer the first option, but for one user attribute (tenant) in my app I'm expecting multiple values. Maybe we should provide both options?

Reply:
I forgot about `getAttributes` method. I could use that in that one case if we only provide the first option.

Reply:
I agree that to get the underlying list you can call `getAttributes().get("tenant")`.

Since there is the possibility of adding both somewhere down the road, I'm leaning more heavily now towards renaming this method to `getFirstAttribute`. I believe this aligns nicely with `MultiValueMap` which is similar in concept and also has a method called `getFirst()`.

Reply:
Hm, after changing `getAttributes` return type, I can't do this anymore
```
List<String> tenants = (List<String>) principal.getAttributes().get("tenant");
```
since the compiler doesn't allow casting from `List<Object>` to `List<String>`.

And this doesn't look very nice: 
```
List<Object> tenants2 = principal.getAttributes().get("tenant");
```

If I had a method like this
```
public <A> List<A> getAttributeValue(String name) {
    return (List<A>) getAttributes().get(name);
}
```
I could do
```
List<String> tenants = principal.getAttributeValue("tenant");
```

Reply:
On the other hand, what if those values have different types... I'm so used to working with only one value type.

Reply:
Good catch! A couple thoughts come to mind:

1. Since this caught you and I off-guard, it would be nice if there was a test in `SimpleSaml2AuthenticatedPrincipalTests` that demonstrated the various behaviors. That way our expectations are codified in a test.
2. I think adding a `List` getter makes sense. An application can set the reference type to `List<Object>` if their attribute contains multiple types.

Also, I'd recommend that the interface stick with `getAttribute` and `getFirstAttribute`. The reason for that is it aligns with similar Spring Security support for OAuth 2.0, which uses `getClaim` and `getAttribute`. As another example, this is the same convention followed by the Servlet spec with `HttpServletRequest#getParameter`, `#getAttribute`, and `#getHeader`.

If you agree, then the contract would become:

```java
Map<String, List<Object>> getAttributes();

<A> List<A> getAttribute(String name);

<A> A getFirstAttribute(String name);
```